### PR TITLE
[1.x] Update Dart client to the updated API of Spine 1.9

### PR DESCRIPTION
**Pull request by armiol**
This changeset updates the Dart client API and implementation in preparation to release Spine 1.9.0 release.

It includes the changes as follows:

**Customization of communication with back-end**

 `HttpTranslator` is introduced to allow customization of HTTP request execution. In particular, it makes possible to define the HTTP headers of the requests sent to back-end, as well as translate the back-end responses. 

Here is how to supply a custom `HttpTranslator`:

```dart

import 'package:spine_client/http_client.dart';

class CustomHttpTranslator extends HttpTranslator {

    // ... 

    @override
    Map<String, String> headers(Uri uri) => {'Custom-Header': 'foo-bar'};
}

// ...
var clients = Clients(backendUrl,
                // ...
                httpTranslator: CustomHttpTranslator(...));
```

If the translator is not customized, the behavior is the same as in previous Dart client releases:
* `application/x-protobuf` content-type set for all requests,
* HTTP request body is a Proto message transformed to bytes, and encoded with Base64,
* HTTP responses are treated as JSON strings, and parsed accordingly. 

**Subscription to entity deletions and archivals**

The recent changes in Spine web back-end made possible to receive the subscription updates whenever the matching entity is archived or deleted. Previously, this functionality was still supported by Dart client API, but in fact it was not functioning properly.

**A word on streams in subscriptions**

The subscription API is build around Dart streams. At the moment, only *broadcast* streams are supported by Dart client. 
While it is possible to use Dart built-in mechanisms to perform the conversion of single-subscription streams into broadcast streams, it cannot be performed with proper performance and behavior generally. Therefore, any custom implementations of `FirebaseClient` should return broadcast streams, performing the conversion internally if needed. 

The related #42 which speaks about this restriction is not addressed in 1.9.0. Instead, it is postponed to future Dart client releases.

**Exposing more API**

Some API which was previously inaccessible to end-users is now available. In particular:

* `json.dart` — the conversion of Proto messages from JSON strings into objects;
* `known_types.dart` — a registry of all Proto types known to the library; required in order to create Proto messages via parsing.

These files were moved from `lib/src` to `lib` folder.

**Dart dependency updates**

Back-merging from `1.8.2` release, this PR removes the dependency onto Firebase client implementation which is specific to web.

The version of Dart libraries is set to `1.9.0-SNAPSHOT.1`.

### CODE REVIEW

**Comment by armiol** [comment]
@dmytro-kashcheiev PTAL.

File: client/lib/client.dart [hunk 1]
```diff
@@ -156,11 +158,20 @@ class Clients {
         }
     }
 
-    static QueryResponseProcessor _chooseProcessor(QueryMode queryMode, FirebaseClient? firebase) {
+    static HttpClient _createHttpClient(String baseUrl, HttpTranslator? httpTranslator) {
+        ArgumentError.checkNotNull(baseUrl, 'baseUrl');
```
Comment (dmytro-kashcheiev) [inline client/lib/client.dart]: Since `baseUrl` in the constructor is not `nullable,` this check is redundant.

File: client/lib/client.dart [hunk 2]
```diff
@@ -156,11 +158,20 @@ class Clients {
...
+        if(httpTranslator == null) {
+            return HttpClient(baseUrl);
+        }
+        return HttpClient.withTranslator(baseUrl, httpTranslator);
+    }
+
+    static QueryProcessor
+    _chooseProcessor(QueryMode queryMode, HttpClient httpClient, FirebaseClient? firebase) {
         ArgumentError.checkNotNull(queryMode, 'queryMode');
```
Comment (dmytro-kashcheiev) [inline client/lib/client.dart]: Same as above.

File: client/lib/http_client.dart [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+class HttpClient {
+
+    final String _baseUrl;
+    final HttpTranslator _translator;
+
+    HttpClient(String baseUrl) : this.withTranslator(baseUrl, HttpTranslator());
+
+    HttpClient.withTranslator(String baseUrl, HttpTranslator translator)
+        : this._translator = ArgumentError.checkNotNull(translator, 'HTTP translator'),
+          this._baseUrl = ArgumentError.checkNotNull(baseUrl, 'base URL');
```
Comment (dmytro-kashcheiev) [inline client/lib/http_client.dart]: Same as above compiler should prevent the passing of `null` here.

File: client/lib/src/query_processor.dart [hunk 4]
```diff
@@ -24,73 +24,68 @@
...
+///
+class FirebaseQueryProcessor implements QueryProcessor {
 
     final FirebaseClient _database;
+    final HttpClient _httpClient;
 
-    FirebaseResponseProcessor(this._database) {
+    FirebaseQueryProcessor(this._database, this._httpClient) {
         ArgumentError.checkNotNull(_database, 'FirebaseClient');
+        ArgumentError.checkNotNull(_httpClient, 'HttpClient');
```
Comment (dmytro-kashcheiev) [inline client/lib/src/query_processor.dart]: Same with `nullability`.

File: integration-tests/client-test/integration-test/reversing_http_translator.dart [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        var reversed = _reverse(responseBody);
+        parseInto(destination, reversed);
+        return destination;
+    }
+
+    static String _reverse(String value) {
+      var reversed = value.split('').reversed.join('');
+      return reversed;
+    }
+}
```
Comment (dmytro-kashcheiev) [inline integration-tests/client-test/integration-test/reversing_http_translator.dart]: Wrong EOF

**Review by dmytro-kashcheiev — APPROVED** [review]
@armiol LGTM with the minor comments. In general, I think `ArgumentError.checkNotNull` in most places is redundant since dart `sounds null safety`.

**Comment by armiol** [comment]
@dmytro-kashcheiev thanks for the comments. I have removed all excessive `null`-related checks.

**Pull request merged**